src/analysis.rs

```diff
@@ -0,0 +1,547 @@
+        // a concatenation, we conservatively assume that the assertion
+        // will pass, so we just drop it. Then we can only get to this
+        // point if we are dealing with some sort of naked empty look.
+        // For now we just do the most conservative thing and say
+        // that such an emptylook could potentially match on any character.
+        &HirKind::Anchor(_) | &HirKind::WordBoundary(_) => FirstSet::anychar(),
+
+        &HirKind::Repetition(ref rep) => {
+            let mut f = fset_of(&*rep.hir);
+            match rep.kind {
+                RepetitionKind::ZeroOrOne => f.accepts_empty = true,
+                RepetitionKind::ZeroOrMore => f.accepts_empty = true,
+                RepetitionKind::OneOrMore => {},
+                RepetitionKind::Range(ref range) => {
+                    match range {
+                        &RepetitionRange::Exactly(0)
+                        | &RepetitionRange::AtLeast(0)
+                        | &RepetitionRange::Bounded(0, _) =>
+                            f.accepts_empty = true,
+                        _ => {}
+                    }
+                }
+            }
+            f
+        },
+        &HirKind::Group(ref group) => fset_of(&group.hir),
+
+        // The most involved case. We need to strip leading empty-looks
+        // as well as take the union of the first sets of the first n+1
+        // expressions where n is the number of leading repetitions.
+        &HirKind::Concat(ref es) => {
+            let mut fset = FirstSet::empty();
+            for (i, e) in es.iter().enumerate() {
+                match e.kind() {
+                    &HirKind::Anchor(_) | &HirKind::WordBoundary(_) => {
+                        // Ignore any leading emptylooks, but any in tail
+                        // position have to be accounted for.
+                        if i == es.len() - 1 {
+                            fset.union(&FirstSet::anychar());
+                        }
+                    }
+                    _ => {
+                        let inner_fset = fset_of(e);
+                        fset.union(&inner_fset);
+
+                        if !inner_fset.accepts_empty() {
+                            // We can stop accumulating after we stop seeing
+                            // first sets which contain epsilon.
+                            // Also, a contatination which terminated by
+                            // one or more expressions which do not accept
+                            // epsilon itself does not acceept epsilon.
+                            fset.accepts_empty = false;
+                            break;
+                        }
+                    }
+                }
+            }
+            fset
+        }
+        &HirKind::Alternation(ref es) => {
+            let mut fset = FirstSet::empty();
+            for e in es {
+                fset.union(&fset_of(e));
```

Comment:
Yeah, this is pretty unforgivable on my part. The solution is definitely going to have to include caching results somewhere, so I've started work rewriting this module as an extra thing that gets constructed while building Hirs like you've suggested elsewhere. I should have it fixed soon.

Reply:
@ethanpailes Awesome! I just want to re-iterate that I am OK with doing structural recursion here for the time being. Literal anaylsis already does this, so it's not introducing a new regression.

The problem that could potentially arise with doing this at HIR construction time is that I think you'll need to keep a copy of fsets for every HIR expression, and that could prove to be quite precarious. This isn't something I realized at the outset when I suggested it. So it's not clear to me that it is _definitely_ the path we want to take.

Sorry for the back and forth on this, but I'm learning as I go. :)

Reply:
I was kinda worried about the additional memory usage, but the fact that FirstSets are represented with a ClassBytes instead of a [u8; 256] does help a bit. What's the use case to worry about for an expanded Hir memory footprint? I would think that there would be more memory used during execution of the regex (at least in the worst case), and I don't think adding a ClassBytes to each Hir node will take up more memory than the 3 Programs and a OnePass that make up an ExecReadOnly. Does the Hir make it into a Regex object somewhere that I havn't noticed?

If I was trigger happy in doing the re-write I can revert that last patch and just try to either add some more ephemeral caching or ditch the visitor to get rid of the quadratic behavior.

Reply:
It's certainly hand wavy, sure, but for example, a naive implementation would add an fset to every character in a pattern. Moreover, the size of an fset is proportion to the size of the fsets of its sub-expressions, so the growth there could be worrisome.

But honestly, I don't know. It's just a concern. This is definitely something I'm going to have to play with.

Looking at your latest commit, I don't think the implementation is actually what we want anyway. If you're looking at sub-sub-expressions _during construction_ (such as with `NestedConcat`), then that's basically wrong and defeats the purpose of computing this property at construction.

Perhaps the best path forward here is to just do the straight-forward thing and use structural recursion. No visitors. No construction time properties. I can sort this out later, perhaps as part of sorting out literal analysis.

Reply:
I think we are saved from storing an fset per character by the Literal hir node, and the size of an fset shouldn't grow any bigger a bit less than 256 bytes. Plus larger fsets should have a smaller memory foot print, unless I'm off the mark on how ClassBytes works.

The NestedConcat issue is a bit trickier. I added NestedConcat while I was working to get the whole test suite to pass. It is only needed for some pretty weird regex, so I can go back to revisit the idea to see if there is a way to get rid of it. (Also isn't strong induction still induction?)

Reply:
> I think we are saved from storing an fset per character by the Literal hir node, and the size of an fset shouldn't grow any bigger a bit less than 256 bytes. Plus larger fsets should have a smaller memory foot print, unless I'm off the mark on how ClassBytes works.

I'm afraid this isn't true. Look at the HIR for a simple expression:

```
$ regex-debug hir 'abc'
Hir {
    kind: Concat(
        [
            Hir {
                kind: Literal(
                    Unicode(
                        'a'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            },
            Hir {
                kind: Literal(
                    Unicode(
                        'b'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            },
            Hir {
                kind: Literal(
                    Unicode(
                        'c'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            }
        ]
    ),
    info: HirInfo {
        bools: 1
    }
}
```

Notice that _every character_ has its own `HirInfo` value. [Your commit adds an fset to `HirInfo`](https://github.com/rust-lang/regex/pull/467/commits/6355ce519b115f1b4c3bcbaac1b824e1e5151309#diff-18cb88398f4c23fb21519633d53f4aefR1658), which means every character will have an fset, which means we have quite a bit of overheard per character. Not only is this problematic in terms of size, but this will actually wind up being a separate independent allocation per character.

Now, this isn't necessarily a death blow to computing the onepass property on construction, but I think it's pretty clear that it requires more work than I thought---potentially even refactoring the HIR itself.

> (Also isn't strong induction still induction?)

Only logically. If you implement induction by actually using every previous step, then you've effectively lost one of the benefits of building this stuff by construction because you still wind up doing work proportional to the depth of the HIR, and that in turn means you've implementing a quadratic algorithm. The really key idea here is that the construction of an HIR expression should _never_ take time proportional to the depth of its sub-expressions. It should always take time proportional to the number of direct sub-expressions.

I'd like to nudge you towards moving back to structural recursion where things are a bit simpler to reason about. I'd also like to say that I don't think we can really permit quadratic algorithms _anywhere_ in this code, unless we explicitly bound them by some small (heuristically chosen) constant.

Reply:
I didn't notice that Literals in Hir only have one character now (I was still thinking of them as the old Expr::Literal). The allocation-per-character thing is an issue. We could solve it by special casing an fset that contains only one byte, but that seems a bit hacky.

I'm going to look into avoiding the use of NestedConcat to fix the worst-case quadratic behavior. I wasn't saying that NestedConcat doesn't make the runtime worst-case quadratic, just quibbling unnecessarily over the word inductive.

Reply:
I don't think you need to fix the use of nested concat because that isn't enough to make this approach work. Could you please go back to the structural recursion implementation as an explicit analysis pass?